lib/instrumentation/http-shared.js

```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
   const auth = opts.headers && (opts.headers.Authorization || opts.headers.authorization)
   return typeof auth === 'string' ? auth.startsWith('AWS4-') : false
 }
+
+// Creates a sanitized URL suitable for the span's HTTP context
+//
+// This function reconstructs a URL using the request object's properties
+// where it can, and falling back to the options where it can not.
+//
+// NOTE: The options argument may not be the same options that are passed
+// to http.request if the caller uses the the http.request(url,options,...)
+// method signature. The agent normalizes the url and options into a single
+// options array. This function expects those pre-normalized options.
+//
+// @param {ClientRequest} req
+// @param {object} options
+// @return string
+function getUrlFromRequestAndOptions (req, options) {
+  if (!req) {
+    return false
+  }
+  options = options || {}
+  req = req || {}
+  req.agent = req.agent || {}
+
+  const port = options.port ? `:${options.port}` : ''
+  const host = req.host || options.hostname || options.host || 'localhost'
+  const protocol = req.protocol || req.agent.protocol
+  let url
+  try {
+    url = new URL(`${protocol}//${host}${port}${req.path}`)
+  } catch (e) {
+    // console.log(e)
+    return false
+  }
+
+  return `${url.protocol}//${url.host}${url.pathname}${url.search}`
```

Comment:
Why the step to `new URL`-parse the values we have, and then reconstruct it to a string again?

Reply:
Per discussion elsewhere, this is to remove the username:password portion of a hostname.  We rely on the URL object's parsing, and then reconstruct the URL to be just the parts we want.